apps/emqx/src/emqx_session.erl

```diff
@@ -328,23 +335,32 @@ add_persistent_subscription(TopicFilterBin, ClientId, Session) ->
 -spec unsubscribe(emqx_types:clientinfo(), emqx_types:topic(), emqx_types:subopts(), session()) ->
     {ok, session()} | {error, emqx_types:reason_code()}.
 unsubscribe(
-    ClientInfo,
+    ClientInfo = #{clientid := ClientId},
     TopicFilter,
     UnSubOpts,
-    Session = #session{subscriptions = Subs}
+    Session0 = #session{subscriptions = Subs}
 ) ->
     case maps:find(TopicFilter, Subs) of
         {ok, SubOpts} ->
             ok = emqx_broker:unsubscribe(TopicFilter),
+            Session1 = remove_persistent_subscription(Session0, TopicFilter, ClientId),
             ok = emqx_hooks:run(
                 'session.unsubscribed',
                 [ClientInfo, TopicFilter, maps:merge(SubOpts, UnSubOpts)]
             ),
-            {ok, Session#session{subscriptions = maps:remove(TopicFilter, Subs)}};
+            {ok, Session1#session{subscriptions = maps:remove(TopicFilter, Subs)}};
         error ->
             {error, ?RC_NO_SUBSCRIPTION_EXISTED}
     end.
 
+-spec remove_persistent_subscription(session(), emqx_types:topic(), emqx_types:clientid()) ->
+    session().
+remove_persistent_subscription(Session, TopicFilterBin, ClientId) ->
+    Iterators = Session#session.iterators,
+    IteratorId = maps:get(TopicFilterBin, Iterators, undefined),
```

Comment:
Q: Why not just `maps:get(TopicFilterBin, Iterators)`? Getting `undefined` here looks kinda impossible.

Reply:
Indeed, I was thinking about idempotency as [here](https://github.com/emqx/emqx/pull/11533#discussion_r1315121515), in a scenario where the iterators are closed but the mnesia entry is not yet removed.

But, on second thought, even in such scenario I think that the ID would still be in the `#session{}` record at this point.

1. Midway while unsubscribing, we close/discard the iterators on all nodes, but the node goes down before `emqx_ds:session_del_iterator` is called.
2. The node is restarted, and the client reconnects.  It'll recreate the same iterators (but starting from `start_time` again) with the same ID.
3. If it tries to unsubscribe again, the iterator IDs will be in the session record, albeit they were reset.

I'll remove this default and the above clause.